test/test.api.js

```diff
@@ -148,13 +148,20 @@ DRIVERS.forEach(function(driverName) {
                             expect(value).to.be(setValue);
 
                             var accumulator = {};
+                            var iterationNumbers = [];
 
-                            localforage.iterate(function(value, key) {
+                            localforage.iterate(function(value, key, iterationNumber) {
                                 accumulator[key] = value;
+                                iterationNumbers.push(iterationNumber);
                             }, function() {
-                                expect(accumulator.officeX).to.be('InitechX');
-                                expect(accumulator.officeY).to.be('InitechY');
-                                done();
+                                try {
```

Comment:
Sorry to bump a really old thread but my comment about why try/catch is necessary was wrong.

It finally dawned on me why these tests are timing out when an expect fails. Mocha is usually good at failing the test due to uncaught exceptions -- I just assumed it wasn't handling it in this case for some reason.

The reason why it's not happening here is because the Error never gets out of the Promise code. LocalForage invokes the callback inside a promise fulfill handler ([localstorage](https://github.com/mozilla/localForage/blob/master/src/drivers/localstorage.js#L296), [indexeddb](https://github.com/mozilla/localForage/blob/master/src/drivers/indexeddb.js#L378), [websql](https://github.com/mozilla/localForage/blob/master/src/drivers/websql.js#L384)), so if the callback throws (i.e. a failed expect) the Error is caught by the promise implementation which then "swallows" the Error and rejects the resulting promise.

try/catch in tests is still useful in this case but only because of how LF invokes callbacks.